lib/jekyll/tags/post_url.rb

```diff
@@ -61,44 +61,86 @@ class PostUrl < Liquid::Tag
       def initialize(tag_name, post, tokens)
         super
         @orig_post = post.strip
-        begin
-          @post = PostComparer.new(@orig_post)
-        rescue StandardError => e
-          raise Jekyll::Errors::PostURLError, <<~MSG
-            Could not parse name of post "#{@orig_post}" in tag 'post_url'.
-             Make sure the post exists and the name is correct.
-             #{e.class}: #{e.message}
-          MSG
-        end
+
+        @template = Liquid::Template.parse(@orig_post) if use_template?
       end
 
       def render(context)
         @context = context
-        site = context.registers[:site]
+        site = @context.registers[:site]
+        orig_post_expanded = expand_orig_post
+
+        begin
+          post = PostComparer.new(orig_post_expanded)
+        rescue StandardError => e
+          raise Jekyll::Errors::PostURLError,
+                could_not_parse_error_message(orig_post_expanded, e)
+        end
 
         site.posts.docs.each do |document|
-          return relative_url(document) if @post == document
+          return relative_url(document) if post == document
         end
 
         # New matching method did not match, fall back to old method
         # with deprecation warning if this matches
 
         site.posts.docs.each do |document|
-          next unless @post.deprecated_equality document
+          next unless post.deprecated_equality document
 
-          Jekyll::Deprecator.deprecation_message(
-            "A call to '{% post_url #{@post.name} %}' did not match a post using the new " \
-            "matching method of checking name (path-date-slug) equality. Please make sure " \
-            "that you change this tag to match the post's name exactly."
-          )
+          Jekyll::Deprecator.deprecation_message(create_deprecation_message(post.name))
           return relative_url(document)
         end
 
-        raise Jekyll::Errors::PostURLError, <<~MSG
-          Could not find post "#{@orig_post}" in tag 'post_url'.
+        raise Jekyll::Errors::PostURLError, could_not_find_error_message(orig_post_expanded)
+      end
+
+      private
+
+      def use_template?
+        content = @orig_post
+        return false if content.nil? || content.empty?
+
+        content.include?("{%") || content.include?("{{")
+      end
```

Comment:
Variable `@orig_post` will always be a string.
So testing for `nil` is redundant.
I'm not on board with allowing `{%` in the markup (because it denotes a Liquid tag which in turn need not always return a string on expansion).
Lastly, if the markup contains `{{` it is definitely not empty. So lets do away with that redundancy.

```ruby
def use_template?
  @orig_post.include?("{{")
end
```
Finally now since this has been distilled to a single line, the method definition itself is unnecessary and can be inlined at point-of-call.